toplevel/opttoploop.ml

```diff
@@ -682,3 +699,23 @@ let run_script ppf name args =
     else name
   in
   use_silently ppf explicit_name
+
+let preprocess_phrase_with_log = preprocess_phrase
+
+let preprocess_phrase ppf phr =
+  let log = Location.init_log ppf in
+  let ans = preprocess_phrase log phr in
+  Log.flush log;
+  ans
+
+let execute_phrase_with_log print_outcome log phr =
+  Fun.protect (fun () -> execute_phrase print_outcome log phr)
+    ~finally:Warnings.reset_fatal
```

Comment:
I do not understand what the `reset_fatal` business is doing here, but this might be due to an inconsistency between toploop and opttoploop (which I am reviewing first).